### ISSUE TITLE
False positive from `library_private_types_in_public_api`

### ISSUE THREAD

**Issue by eernstg** [dart-lang/linter#4858]
Consider the following program:

```dart
// ----- Library 'public.dart'.

class _Private {
  final int i;
  const _Private(this.i);
}

extension type const Public._(_Private it) implements _Private { // LINT            <----
  const Public.small() : this._(const _Private(10));
  const Public.large() : this._(const _Private(100));
  factory Public.negative(int i) {
    if (i >= 0) throw ArgumentError("Must be negative");
    return Public._(_Private(i));
  }
}

// ----- Some other library
import 'public.dart';

// Can create instances of `_Private` using the "public face" `Public`,
// but only in very specific ways.
// Can use `Public` as the static type of instances of `_Private`.

void main() {
  const c1 = Public.small();
  const c2 = Public.large();
  var v1 = Public.negative(-1);
  Public v2 = c1;
  print(v2.i); // '10'.
}
```

The program gives rise to a diagnostic message from `library_private_types_in_public_api` at the formal parameter `it` of the primary constructor `Public._`:

```
Invalid use of a private type in a public API.
```

However, a private constructor should probably not be considered to be a public API.

Also, the overall purpose of the code seems legitimate (that is, providing access for other libraries to a private class as a type, and providing controlled access to the creation of instances of the private class), which could serve as an additional argument why the lint should not flag this kind of situation.

**Comment by pq** [comment]
@bwilkerson

**Comment by bwilkerson** [comment]
> However, a private constructor should probably not be considered to be a public API.

I agree. A private constructor is not public API.

However, as I understand it, the representation field (`_Private it`) is treated like a final instance field with the same name and type. If the name of the field is not private, then the induced getter is public API with a private return type. I believe that that's the public API being reported by the lint, not the constructor.

> ... providing access for other libraries to a private class as a type ...

I believe that's exactly the kind of thing the lint is designed to prevent. If you really want the other libraries to use instances of a class then that class shouldn't be private.

**Comment by eernstg** [comment]
> the representation field `(_Private it)` is treated like a final instance field

Aha, that was the reason. Good catch, thanks!

I just changed it to `_it` (which wasn't important for the example), and the lint message went away.

OK, so this entire issue was a false positive. ;-)
Closing.

**Comment by bwilkerson** [comment]
Ok. I don't have any problem with your example [1]. If you added something like
```
extension type const Public._(C it) implements C {}
```
everything would be fine because `C` is a public type.

But if you wrote
```
extension type const Public._(_C it) implements _C {}
```
I'd expect the lint to produce a diagnostic because then you're exposing the private type instead of the public face you defined.

[1] I never would have thought of an example like that, hence my not mentioning it as a possible solution. Also, wouldn't this be better supported if we just implemented the variance language feature rather than require code like this?
